python/tile_zarr_base.py

```diff
@@ -0,0 +1,76 @@
+#!/usr/bin/env python3
+
+import numpy as np
+import dask.array as da
+from numcodecs import Zlib
+
+from pathlib import Path
+import argparse
+
+
+def tile_zarr(
+    zarr_pyramid_base, max_level=None, compressor=None, dtype=None,
+):
+    TILE_SIZE = 512
+    img = da.from_zarr(zarr_pyramid_base)
+    pyramid_path = Path(zarr_pyramid_base).parent
+    chunks = img.chunksize
+
+    if dtype is None:
+        dtype = img.dtype
+    if max_level is None:
+        # create all levels up to 512 x 512
+        max_level = (
+            int(np.ceil(np.log2(np.maximum(img.shape[1], img.shape[2])))) - 9
+        )
+    if compressor is None:
+        compressor = Zlib(level=1)
+
+    for i in range(1, max_level):
+        img = da.coarsen(np.mean, img, {1: 2, 2: 2}, trim_excess=True)
+
+        # Edge Case: Need to pad smallest thumbnail sometimes.
+        #
+        # If a dimension of an array is larger than TILE_SIZE,
+        # zarr will respect the chunk size requested an automatically
+        # pad with zeros in the store. However, if an array dimension
+        # is smaller than the tile size, `da.to_zarr` will change the
+        # chunking and not pad with zeros. We need sometimes need to pad
+        # for the smallest tiles because x and y might not be square.
+        if img.shape[1] < TILE_SIZE:
+            img = da.pad(
+                img,
+                ((0, 0), (0, TILE_SIZE - img.shape[1]), (0, 0)),
+                "constant",
+            )
+
+        if img.shape[2] < TILE_SIZE:
+            img = da.pad(
+                img,
+                ((0, 0), (0, 0), (0, TILE_SIZE - img.shape[2])),
+                "constant",
+            )
+
+        # Define pyramid level path
+        out_path = str(pyramid_path / str(i).zfill(2))
```

Comment:
I think I'd prefer if our `_path` variables were Paths.

Reply:
Unfortunately zarr and dask.array don't accept `Path` objects. I use pathlib to ensure that the paths can be cross platform, and I believe the `Path.__str__`takes care of creating the write path for the file system.

Reply:
@mccalluc To clarify, `pyramid_path` is a Path but `out_path` must be a string for zarr and dask array. I could use `str(out_path)` in the `da.to_zarr` and `da.from_zarr` if that is preferable.